cli/open.go

```diff
@@ -0,0 +1,268 @@
+				if err != nil {
+					return xerrors.Errorf("create API key: %w", err)
+				}
+				qp.Add("token", apiKey.Key)
+			}
+
+			u.RawQuery = qp.Encode()
+
+			openingPath := workspaceName
+			if directory != "" {
+				openingPath += ":" + directory
+			}
+
+			if insideAWorkspace {
+				_, _ = fmt.Fprintf(inv.Stderr, "Opening %s in %s is not supported inside a workspace, please open the following URI on your local machine instead:\n\n", openingPath, vscodeDesktopName)
+				_, _ = fmt.Fprintf(inv.Stdout, "%s\n", u.String())
+				return nil
+			}
+			_, _ = fmt.Fprintf(inv.Stderr, "Opening %s in %s\n", openingPath, vscodeDesktopName)
+
+			if !testOpenError {
+				err = open.Run(u.String())
+			} else {
+				err = xerrors.New("test.open-error")
+			}
+			if err != nil {
+				if !generateToken {
+					qp.Del("token")
+					u.RawQuery = qp.Encode()
+				}
+
+				_, _ = fmt.Fprintf(inv.Stderr, "Could not automatically open %s in %s: %s\n", openingPath, vscodeDesktopName, err)
+				_, _ = fmt.Fprintf(inv.Stderr, "Please open the following URI instead:\n\n")
+				_, _ = fmt.Fprintf(inv.Stdout, "%s\n", u.String())
+				return nil
+			}
+
+			return nil
+		},
+	}
+
+	cmd.Options = clibase.OptionSet{
+		{
+			Flag: "generate-token",
+			Env:  "CODER_OPEN_VSCODE_GENERATE_TOKEN",
+			Description: fmt.Sprintf(
+				"Generate an auth token and include it in the vscode:// URI. This is for automagical configuration of %s and not needed if already configured. "+
+					"This flag does not need to be specified when running this command on a local machine unless automatic open fails.",
+				vscodeDesktopName,
+			),
+			Value: clibase.BoolOf(&generateToken),
+		},
+		{
+			Flag:        "test.open-error",
+			Description: "Don't run the open command.",
+			Value:       clibase.BoolOf(&testOpenError),
+			Hidden:      true, // This is for testing!
+		},
+	}
+
+	return cmd
+}
+
+// isWindowsAbsPath checks if the path is an absolute path on Windows. On Unix
+// systems the check is very simplistic and does not cover edge cases.
+//
+//nolint:revive // Shadow path variable for readability.
+func isWindowsAbsPath(path string) bool {
+	if runtime.GOOS == "windows" {
+		return filepath.IsAbs(path)
+	}
+
+	switch {
+	case len(path) >= 2 && path[1] == ':':
+		// Path starts with a drive letter.
+		return len(path) == 2 || (len(path) >= 4 && path[2] == '\\' && path[3] == '\\')
+	default:
+		return false
+	}
+}
+
+// windowsJoinPath joins the elements into a path, using Windows path separator
+// and converting forward slashes to backslashes. On Unix systems a very
+// simplistic join operator is used.
+func windowsJoinPath(elem ...string) string {
+	if runtime.GOOS == "windows" {
+		return filepath.Join(elem...)
+	}
+
+	var s string
+	for _, e := range elem {
+		e = strings.ReplaceAll(e, "/", "\\")
+		if e == "" {
+			continue
+		}
+		if s == "" {
+			s = e
+			continue
+		}
+		s += "\\" + strings.TrimSuffix(s, "\\")
+	}
+	return s
+}
```

Comment:
I think these are ripe to be harvested into `cliutil` :-)

Reply:
I'd like to defer this move for now since I'm not convinced of their utility outside the scope of `open` (yet). I've also kept the `resolveAgentAbsPath` and added an internal test for it, which also calls these methods.

Ultimately I'd like something like the `path/filepath` package, where you can set the `Separator` yourself, but I also don't feel like copying over all that code for now. 😅

Reply:
Fair, we can do this later. 👍